.travis.yml

```diff
@@ -31,32 +36,84 @@ cache:
-  # "copyRuntimeLibs" copies the JAR *and* dependencies - which Freetalk also
-  # needs - to build/output/
-  - ./gradlew jar copyRuntimeLibs -x test
+  - git reset --hard # Ensure accidental changes don't persist due to the Travis cache.
+  - git fetch && if [ "$(git rev-parse @)" != "$(git rev-parse @{u})" ] ; then FRED_UPDATED=1 ; git pull ; fi
   - cd "$TRAVIS_BUILD_DIR"
-  # Print the checksums of the Freetalk dependencies - for debugging
-  - sha256sum ../fred/build/output/*
 
-script: ant
+# Compile Freenet Git repository
+before_script: |
+  if [ "$FRED_UPDATED" = 1 ] ; then
+    cd "$TRAVIS_BUILD_DIR"/../fred &&
+    # TODO: freenet.jar won't contain class Version if we don't run the
+    # clean task in a separate execution of Gradle. I.e. this wouldn't work:
+    #   $ gradle clean jar
+    # This is due to a bug in fred's Gradle script which could be fixed
+    # like this WoT commit did: 06c007204f40c712a398f0b58671f77fd9aeffd1
+    ./gradlew clean &&
+    # "copyRuntimeLibs" copies the JAR *and* dependencies - which Freetalk also
+    # needs - to build/output/
+    ./gradlew jar copyRuntimeLibs -x test &&
+    cd "$TRAVIS_BUILD_DIR"
+  else
+    echo "No changes at fred, not recompiling."
+  fi
+
+# Compile and test Freetalk
+script:
+  - set -o errexit
+  - echo 'Checksums of dependencies:' ; sha256sum ../fred/build/output/*
+  # Don't allow Travis to override the low memory limit which our builder sets with a higher one.
+  - unset _JAVA_OPTIONS
+  - echo 'fred version:' ; java -classpath '../fred/build/output/freenet.jar' 'freenet.node.Version'
+  - rm -f ./gradlew
+  - |
+    if [ "$(tr -cd '[:digit:]' <<< "$TRAVIS_JDK_VERSION")" -lt 13 ] ; then # Java versions < 13
+      # Latest Gradle version as of 2021-01-22.
+      gradle wrapper --gradle-version 6.8.1
+    else
+      # As of 2020-08-06 "gradle wrapper" fails on Java 13 and above so use the one of fred instead
+      # as a workaround.
+      # TODO: Code quality: Remove this workaround once the issue is fixed.
+      ln -s "$TRAVIS_BUILD_DIR"/../fred/gradlew ./gradlew
+    fi
+  - ./gradlew --version
+  # To test the Ant and Gradle builders against each other uncomment the following.
+  # (The scripts will use the ./gradlew we've installed so we can ensure the latest Gradle works.)
+  # FIXME: Enabled to test the currently in development Gradle changes. Disable afterwards.
+  - tools/compare-gradle-jars-with-ant-jars
+  - tools/compare-gradle-tests-with-ant-tests
+  # Show stdout/stderr so random seeds of failed tests can be obtained by developers to
+  # reproduce failed test runs. Also prevents the 10 minute build timeout.
+  - FREETALK__SHOW_GRADLE_TEST_OUTPUT=1 ./gradlew clean test jar
 
 jdk:
-  - oraclejdk9
-  - oraclejdk8
+  # openjdk7: Not available anymore.
   - openjdk8
-  - openjdk7
-  # openjdk9: As of 2018-05-12 isn't available on Travis yet.
+  - openjdk9
+  - openjdk10
+  - openjdk11
+  - openjdk12
+  - openjdk13
+  - openjdk14
+  - openjdk15
```

Comment:
For Freenet I now switched to the minimum set that includes the most likely installations: openjdk8, openjdk11 (latest LTS), and openjdk15 (latest).